tests/php/test-class-amp-dom-document.php

```diff
@@ -0,0 +1,328 @@
+				<td class="text" style="width:10%;">1947</td>
+				<td class="text">Pittsburgh Ironmen</td>
+				<td class="boolean" style="width:10%;text-align:center;"></td>
+				<td class="number" style="width:10%;">1242</td>
+				<td class="number">1192</td>
+				<td class="number">1111</td>
+				<td class="number highlight">1182</td>
+				</tr>
+			';
+		}
+
+		$html .= '</table></body></html>';
+
+		$to_convert = sprintf(
+			'<amp-iframe sandbox="allow-scripts" srcdoc="%s"> </amp-iframe>',
+			htmlentities( $html )
+		);
+
+		Document::from_html( $to_convert )->saveHTML();
+
+		$this->assertSame( PREG_NO_ERROR, preg_last_error(), 'Probably failed when backtrack limit was exhausted.' );
+	}
+
+	/**
+	 * Test that HEAD and BODY elements are always present.
+	 *
+	 * @covers \Amp\AmpWP\Dom\Document::normalize_document_structure()
+	 */
+	public function test_ensuring_head_body() {
+		// The meta charset tag that is automatically added needs to always be taken into account.
+
+		$html = '<html><body><p>Hello</p></body></html>';
+		$dom  = Document::from_html( $html );
+		$this->assertEquals( 'head', $dom->documentElement->firstChild->nodeName );
+		$this->assertEquals( 1, $dom->head->childNodes->length );
+		$this->assertEquals( 'body', $dom->documentElement->lastChild->nodeName );
+		$this->assertEquals( $dom->body, $dom->getElementsByTagName( 'p' )->item( 0 )->parentNode );
+
+		$html = '<html><head><title>foo</title></head></html>';
+		$dom  = Document::from_html( $html );
+		$this->assertEquals( 'head', $dom->documentElement->firstChild->nodeName );
+		$this->assertEquals( $dom->head->firstChild, $dom->getElementsByTagName( 'meta' )->item( 0 ) );
+		$this->assertEquals( $dom->head->firstChild->nextSibling, $dom->getElementsByTagName( 'title' )->item( 0 ) );
+		$this->assertEquals( 'body', $dom->documentElement->lastChild->nodeName );
+		$this->assertEquals( 0, $dom->body->childNodes->length );
+
+		$html = '<html><head><title>foo</title></head><p>no body</p></html>';
+		$dom  = Document::from_html( $html );
+		$this->assertEquals( 'head', $dom->documentElement->firstChild->nodeName );
+		$this->assertEquals( $dom->head->firstChild, $dom->getElementsByTagName( 'meta' )->item( 0 ) );
+		$this->assertEquals( $dom->head->firstChild->nextSibling, $dom->getElementsByTagName( 'title' )->item( 0 ) );
+		$p = $dom->getElementsByTagName( 'p' )->item( 0 );
+		$this->assertEquals( $dom->body, $p->parentNode );
+		$this->assertEquals( 'no body', $p->textContent );
+
+		$html = 'Hello world';
+		$dom  = Document::from_html( $html );
+		$this->assertEquals( 'head', $dom->documentElement->firstChild->nodeName );
+		$this->assertEquals( 1, $dom->head->childNodes->length );
+		$this->assertEquals( 'body', $dom->documentElement->lastChild->nodeName );
+		$this->assertEquals( 'Hello world', $dom->body->lastChild->textContent );
+	}
+
+
+	/**
+	 * Test that invalid head nodes are moved to body.
+	 *
+	 * @covers \Amp\AmpWP\Dom\Document::move_invalid_head_nodes_to_body()
```

Comment:
This is fully-qualified but `Document::loadHTML()` above is not. Intentional?

Reply:
Shouldn't matter, but I'll make them consistent.

When using namespace, I prefer to stick with omitting the root namespace, so I'll go with that.

Reply:
Oh, our tests don't have namespaces, so I guess I can only still with the FQCN anyway.

Reply:
Ah, we're importing Document, so just using `Document` should be fine, and causes less visual clutter.

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/3758/commits/7c037a94f72b8e3ea07b2a60d3ad8179f9740da7